// rsuite-table@5.2.0/lib/Loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Loader = /*#__PURE__*/_react["default"].forwardRef(function (props, ref) {
  var loadAnimation = props.loadAnimation,
      loading = props.loading,
      locale = props.locale,
      addPrefix = props.addPrefix,
      renderLoading = props.renderLoading;

  if (!loadAnimation && !loading) {
    return null;
  }

  var loadingElement = /*#__PURE__*/_react["default"].createElement("div", {
    ref: ref,
    className: addPrefix('loader-wrapper')
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: addPrefix('loader')
  }, /*#__PURE__*/_react["default"].createElement("i", {
    className: addPrefix('loader-icon')
  }), /*#__PURE__*/_react["default"].createElement("span", {
    className: addPrefix('loader-text')
  }, locale === null || locale === void 0 ? void 0 : locale.loading)));

  return renderLoading ? renderLoading(loadingElement) : loadingElement;
});

Loader.propTypes = {
  locale: _propTypes["default"].object,
  loadAnimation: _propTypes["default"].bool,
  loading: _propTypes["default"].bool,
  addPrefix: _propTypes["default"].func,
  renderLoading: _propTypes["default"].func
};
Loader.displayName = 'Table.Loader';
var _default = Loader;
exports["default"] = _default;
